File: CCC/2025/ccc25j5.py
# CCC '25 J5 - Connecting Territories
import sys
input = sys.stdin.readline


R,C,M = int(input()),int(input()),int(input())
if R*C <= M:
    res = 0
    for r in range(R):
        res += (r*C) % M + 1
    print(res)
    sys.exit(0)

# orz bruce 
pre = [0] * C
for r in range(R):
    cur = [0] * C
    for c in range(C):
        cell = (r*C+c) % M + 1
        if c == 0:
            cur[c] = min(pre[c],pre[c+1]) + cell
        elif c < C-1:
            cur[c] = min(pre[c-1],pre[c],pre[c+1]) + cell
        else:
            cur[c] = min(pre[c-1],pre[c]) + cell
        
    pre,cur = cur,pre


print(min(pre))

# orz bruce
# R, C, M = int(input()), int(input()), int(input())
# pre = [0] * C
# for i in range(R):
#    cur = [0] * C
#    for j in range(C):
#        cell = (i * C + j) % M + 1
#        if j == 0:
#            cur[j] = min(pre[j], pre[j+1]) + cell
#        elif j < C-1:
#            cur[j] = min(pre[j-1], pre[j], pre[j+1]) + cell
#        else:
#            cur[j] = min(pre[j-1], pre[j]) + cell
#    pre, cur = cur, pre
# print(min(pre))

File: DMOJ/CCC/2025/ccc25j5.cpp
// CCC '25 J5 - Connecting Territories
#include <bits/stdc++.h>
using namespace std;


int main() {
    ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);

    int R,C,M; cin >> R >> C >> M;
    if ((R*C) <= M) {
        long long res = 0;
        for (int r = 0; r < R; r++) {
            res += (r*C) % M + 1;
        }
        cout << res << "\n"; return 0;
    }

    // orz bruce
    vector<long long> pre(C,0);
    for (int r = 0; r < R; r++) {
        vector<long long> cur(C,0);
        for (int c = 0; c < C; c++) {
            long long cell = (r*C+c) % M + 1;
            if (c == 0) {
                cur[c] = min(pre[c],pre[c+1]) + cell;
            }
            else if (c < C-1) {
                cur[c] = min({pre[c-1],pre[c],pre[c+1]}) + cell;
            }
            else {
                cur[c] = min(pre[c-1],pre[c]) + cell;
            }
        }
        swap(pre,cur);
    }
    
    cout << *min_element(pre.begin(),pre.end()) << "\n";

    return 0;
}
